### PR TITLE
WIP

### PR DESCRIPTION
**Pull request by EvanPiro**

### CODE REVIEW

File: src/Portfolio.sol [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+contract Portfolio {
+    function uniswapV3(
+        uint256 amountIn,
+        address tokenIn,
+        address tokenOut
+    ) internal returns (uint256 amountOut) {
+        address router = 0xE592427A0AEce92De3Edee1F18E0157C05861564;
+        uint24 fee = 3000;
+        address WETH = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
```
Comment (EvanPiro) [inline src/Portfolio.sol:22]: These addresses should be passed through the constructor so that they can be network agnostic.

File: src/Coin.sol [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    function mint(
+        address _positionAddress,
+        address _receiver,
+        uint256 _moreDebt
+    ) external {
+        // require(
+        //     notary.isValidPosition(_positionAddress),
+        //     "Caller is not authorized to mint"
+        // );
+        _mint(_receiver, _moreDebt);
```
Comment (EvanPiro) [inline src/Coin.sol:39]: Anyone can call this minting function.

File: src/Coin.sol [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    }
```
Comment (EvanPiro) [inline src/Coin.sol:40]: Anyone can call this minting function. Perhaps the following could fix it:

```suggestion
    /**
     * @dev Mints for authenticated position contracts.
     */
    function mint(address _receiver, uint256 _moreDebt) public {
        address _positionAddress = msg.sender;
        require(notary.isValidPosition(_positionAddress), "Caller is not authorized to mint");
        Position position = Position(_positionAddress);
        require(position.canTake(_moreDebt), "Position cannot take debt");
        _mint(_receiver, _moreDebt);
    }
```

File: src/Position.sol [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    Coin private immutable coin;
+    uint256 public debt;
+    bool isInsolvent;
+    Basket collateral;
+    uint256 public constant RATIO = 150;
+    uint256 public constant COLLATERAL_DECIMAL = 1e18;
+    address user;
+    uint256 public constant FEE = 500;
+    uint256 public constant PENALTY = 500;
+    uint256 private s_lastTimeStamp;
```
Comment (EvanPiro) [inline src/Position.sol:32]: what is `s_lastTimeStamp`? It's hard to infer from the name.

File: src/Position.sol [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        require(canTake(_moreDebt), "Position cannot take debt");
+        coin.mint(address(this), _receiver, _moreDebt);
+        debt = debt + _moreDebt;
+    }
+
+    /**
+     * @dev Liquidates vault if priceId points to a price record of insolvency.
+     * This function should be in the notary contract with regular checks
+     */
+    function liquidate(uint256 priceId) public {
```
Comment (EvanPiro) [inline src/Position.sol:102]: This could use a large comment above explaining the algorithm of the liquidation.

File: src/Portfolio.sol (hunk 1 above)
Comment (MMtis) [reply src/Portfolio.sol:22]: Agreed.

File: src/Position.sol (hunk 4 above)
Comment (MMtis) [reply src/Position.sol:32]: lastTimeStamp is a variable that store the time (or block timestamp). This variable will be used in the calculation of accrued interest.

File: src/Coin.sol (hunk 3 above)
Comment (MMtis) [reply src/Coin.sol:40]: Agreed, only the Notary (or manager) and the position creator (debtor) should be able to mint. Notary should have all access and debtor will only be able to mint an amount depending on the collateral deposited.

File: src/Notary.sol [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            user,
+            address(this),
+            portfolioAddress,
+            ethusd
+        );
+        address _vaultAddress = address(vault);
+
+        isValidPosition[_vaultAddress] = true;
+        vaults.push(vault);
+        vaultID += 1;
```
Comment (EvanPiro) [inline src/Notary.sol:72]: Maintaining a list all vaults should be the responsibility of off-chain services, not this contract. Only a mapping of vault addresses is sufficient to notarizing vaults.

File: src/Notary.sol [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        emit VaultOpened(_vaultAddress);
+        return _vaultAddress;
+    }
+
+    function liquidateVaults() public onlyOwner {
+        uint256 length = vaults.length;
+        for (uint256 i = 0; i < length; i++) {
+            vaults[i].liquidate();
+        }
+    }
```
Comment (EvanPiro) [inline src/Notary.sol:83]: This function might be better in a liquidator contract that isn't core to the protocol. How liquidations are aggregated is the responsibility of the liquidating party.

File: src/Vault.sol [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+        _;
+    }
+
+    constructor(
+        address _coinAddress,
+        address _user,
+        address _notary,
+        address _portfolio,
+        address _EthUSD
+    ) {
```
Comment (EvanPiro) [inline src/Vault.sol:71]: If we were to use the [`Ownable`](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/access/Ownable.sol) to implement the vault ownership, we would get ownership transferability out of the box.

Comment (MMtis) [reply src/Vault.sol:71]: Agreed. I'm currently making the changes.

File: src/Notary.sol (hunk 6 above)
Comment (MMtis) [reply src/Notary.sol:72]: Why ?

Comment (EvanPiro) [reply src/Notary.sol:72]: Just to eliminate the amount of on-chain storage and computation a list requires. What's the use case of iterating through all vaults on-chain?

**Pull request merged**